Phase08/Phase08/PorterStemmer.cs

```diff
@@ -0,0 +1,297 @@
+        {
+            string letterTypes = GetLetterTypes(input);
+            int m = GetM(letterTypes);
+            if (m > 1 && input.EndsWith("ll"))
+            {
+                return RemoveLastCharacters(input, 1);
+            }
+            return input;
+        }
+
+        private char GetLastDoubleConsonant(string input)
+        {
+            if (input.Length < 2) return (char)0;
+            char lastLetter = input[input.Length - 1];
+            char penultimateLetter = input[input.Length - 2];
+            if (lastLetter == penultimateLetter && GetLetterType((char)0, lastLetter) == 'c')
+            {
+                return lastLetter;
+            }
+            return (char)0;
+        }
+
+        private bool IsStarO(string input)
+        {
+            if (input.Length < 3) return false;
+
+            char lastLetter = input[input.Length - 1];
+            if (lastLetter == 'w' || lastLetter == 'x' || lastLetter == 'y') return false;
+
+            char secondToLastLetter = input[input.Length - 2];
+            char thirdToLastLetter = input[input.Length - 3];
+            char fourthToLastLetter = (char)(input.Length == 3 ? 0 : input[input.Length - 4]);
+            return GetLetterType(secondToLastLetter, lastLetter) == 'c'
+                    && GetLetterType(thirdToLastLetter, secondToLastLetter) == 'v'
+                    && GetLetterType(fourthToLastLetter, thirdToLastLetter) == 'c';
+        }
+
+        private string GetLetterTypes(string input)
+        {
+            StringBuilder letterTypes = new StringBuilder(input.Length);
+            for (int i = 0; i < input.Length; i++)
+            {
+                char letter = input[i];
+                char previousLetter = (char)(i == 0 ? 0 : input[i - 1]);
+                char letterType = GetLetterType(previousLetter, letter);
+                if (letterTypes.Length == 0 || letterTypes[letterTypes.Length - 1] != letterType)
+                {
+                    letterTypes.Append(letterType);
+                }
+            }
+            return letterTypes.ToString();
+        }
+
+        int GetM(string letterTypes)
+        {
+            if (letterTypes.Length < 2) return 0;
+            if (letterTypes[0] == 'c') return (letterTypes.Length - 1) / 2;
+            return letterTypes.Length / 2;
+        }
+
+        private char GetLetterType(char previousLetter, char letter)
+        {
+            switch (letter)
+            {
+                case 'a':
+                case 'e':
+                case 'i':
+                case 'o':
+                case 'u':
+                    return 'v';
+                case 'y':
+                    if (previousLetter == 0 || GetLetterType((char)0, previousLetter) == 'v')
+                    {
+                        return 'c';
+                    }
+                    return 'v';
+                default:
+                    return 'c';
+            }
+        }
+    }
+}
+
+    
```

Comment:
nice stemming but it would be better if u use the chain of responsibility design pattern . it will make refactoring and developing much easier